# HG changeset patch
# User Maxim Dounin <mdounin@mdounin.ru>
# Date 1315324342 -14400
# Node ID d603ce98fada855f0100b422b7b5672fd22fabea
# Parent  69e2e11aa425663063801a333a29b305593a6925
Better handling of late upstream creation.
Configuration with duplicate upstream blocks defined after first use, i.e.
like
    server {
        ...
        location / {
            proxy_pass http://backend;
        }
    }
    upstream backend { ... }
    upstream backend { ... }
now correctly results in "duplicate upstream" error.
Additionally, upstream blocks defined after first use now handle various
server directive parameters ("weight", "max_fails", etc.).  Previously
configuration like
    server {
        ...
        location / {
            proxy_pass http://backend;
        }
    }
    upstream backend {
        server 127.0.0.1 max_fails=5;
    }
incorrectly resulted in "invalid parameter "max_fails=5"" error.
diff --git a/src/http/ngx_http_upstream.c b/src/http/ngx_http_upstream.c
--- a/src/http/ngx_http_upstream.c
+++ b/src/http/ngx_http_upstream.c
@@ -4256,6 +4256,10 @@ ngx_http_upstream_add(ngx_conf_t *cf, ng
             continue;
         }
 
+        if (flags & NGX_HTTP_UPSTREAM_CREATE) {
+            uscfp[i]->flags = flags;
+        }
+
         return uscfp[i];
     }
 
